feat: debug options

pierrezimmermannbam opened the pull request:
<!-- Please provide enough information so that others can review your pull request. -->
<!-- Keep pull requests small and focused on a single change. -->

### Summary

Adds optional filterProps option for debug util as discussed in #1135
<!-- What existing problem does the pull request solve? Can you solve the issue with a different approach? -->

### Test plan

I added some tests using the option and filtering through propName, propValue or node
Additionally I checked that filtering props in debug doesn't mutate initial instance

<!-- List the steps with which we can test this change. Provide screenshots if this changes anything visual. -->

AugustinLF commented on src/__tests__/render.test.tsx:
```diff
@@ -191,11 +195,34 @@ test('debug changing component', () => {
+
+  const { debug, getByText } = render(<Banana />);
+
+  debug({ filterProps: (propName) => propName === 'children' });
```
I'd either break those in different tests (to have better snapshot tests), and/or inline the snapshots. The snapshot are very "isolated" right now

AugustinLF commented on src/render.tsx:
```diff
@@ -135,18 +135,18 @@ function updateWithAct(
 
 interface DebugFunction {
-  (message?: string): void;
+  (options?: string | DebugOptions): void;
```
Please update flow types (I wish we could drop them^^')

pierrezimmermannbam replied:
Oh I forgot about these, I updated them

AugustinLF reviewed (changes requested):
Looks pretty good overall (even though I'm not familiar with the way `pretty-format` works, two additional points:
- can we get `mapNode` also as an option?
- can you try adding this option to the global `configure` options? I believe there is a lot of values in setting a value for a whole project (I'll definitely filter out all my `style` props)

pierrezimmermannbam commented:
@AugustinLF I added the option to configure the debugOptions globally and treated your other comments. Also I have updated the website docs with the new api. Regarding the mapNode option I was considering adding it in another pr if that's ok with you, so that the review process is shorter and also because I think this option by itself makes sense so the faster it is available the better

mdjastrzebski commented on package.json:
```diff
@@ -97,6 +97,7 @@
       "/node_modules/(?!(@react-native|react-native)/).*/"
-    ]
+    ],
+    "clearMocks": true
```
Is this change related to his PR? Why is it needed?

pierrezimmermannbam replied:
It's not really related to this pr but I noticed that mocks weren't automatically cleared because I had to add tests on debug and since the calls to console.log were not cleared, my tests did not pass after I ran them with the other ones. I think tests should all pass when they are run by themselves and should not depend on other tests and automatically clearing mocks call history helps achieving that

mdjastrzebski commented on src/config.ts:
```diff
@@ -1,6 +1,9 @@
 export type Config = {
   /** Default timeout, in ms, for `waitFor` and `findBy*` queries. */
   asyncUtilTimeout: number;
+  debugOptions?: Partial<DebugOptions>;
```
RTL/DTL convention seems to be [`defaultXxx`](https://testing-library.com/docs/dom-testing-library/api-configuration#defaulthidden)
```suggestion
  defaultDebugOptions?: Partial<DebugOptions>;
```

mdjastrzebski commented on src/helpers/format.ts:
```diff
@@ -1,10 +1,58 @@
     highlight: true,
   });
 
+const getCustomPlugin = (filterProps?: FilterNodeByTypeFunction): NewPlugin => {
```
```suggestion
const reactTestComponentPlugin = (filterProps?: FilterNodeByTypeFunction): NewPlugin => {
```

mdjastrzebski commented on src/helpers/format.ts:
```diff
@@ -1,10 +1,58 @@
+) =>
   prettyFormat(input, {
-    plugins: [plugins.ReactTestComponent, plugins.ReactElement],
+    plugins: [getCustomPlugin(options.filterProps)],
```
Seems like we might as well use the orginal ReactElement plugin and just tweak the ReactTestComponent
```suggestion
    plugins: [reactTestComponentPlugin(options.filterProps), plugins.ReactElement],
```

mdjastrzebski commented on src/helpers/format.ts:
```diff
@@ -1,10 +1,58 @@
+  return {
+    test: (val) =>
+      plugins.ReactTestComponent.test(val) || plugins.ReactElement.test(val),
+    serialize: (val, config, indentation, depth, refs, printer) => {
```
As mentioned above you should be able to make this impl. easier by just focusing on `ReactTestComponent` part and using `plugins.ReactElement` as-is.

mdjastrzebski commented on typings/index.flow.js:
```diff
@@ -306,8 +306,22 @@ interface Thenable {
+  node: ReactTestRendererJSON
+) => boolean;
+
+type FormatOptions = {
+  filterProps?: FilterNodeByTypeFunction,
+};
+
+type DebugOptions = {
+  message?: string,
+} & FormatOptions;
```
nit: I would go with interface here, but this is purely personal style pref.

```suggestion
interface FormatOptions {
  filterProps?: FilterNodeByTypeFunction,
};

interface DebugOptions extends FormatOptions {
  message?: string,
}
```

mdjastrzebski commented on typings/index.flow.js:
```diff
@@ -306,8 +306,22 @@ interface Thenable {
+
 type Debug = {
-  (message?: string): void,
+  (options?: string | DebugOptions): void,
```
```suggestion
  (options?: DebugOptions | string): void,
```

mdjastrzebski commented on website/docs/API.md:
```diff
@@ -167,15 +168,24 @@ Usually you should not need to call `unmount` as it is done automatically if you
+  filterProps?: FilterNodeByTypeFunction;
+}
+
+debug(options?: string | DebugOptions): void
```
```suggestion
debug(options?: DebugOptions | string): void
```

mdjastrzebski commented on website/docs/API.md:
````diff
@@ -167,15 +168,24 @@ Usually you should not need to call `unmount` as it is done automatically if you
 ```jsx
 render(<Component />);
 
 screen.debug('optional message');
````
Let's suggest them only the modern way.
```suggestion
```

mdjastrzebski commented on website/docs/API.md:
```diff
@@ -167,15 +168,24 @@ Usually you should not need to call `unmount` as it is done automatically if you
 render(<Component />);
 
 screen.debug('optional message');
+screen.debug({message: 'optional message'});
```
formatting:
```suggestion
screen.debug({ message: 'optional message' });
```

mdjastrzebski commented on website/docs/API.md:
````diff
@@ -190,6 +200,16 @@ optional message
+debug({filterProps : (propName, propValue, node) => propName !== 'style'})
+```
+
+This will log the rendered JSX without style props
````
```suggestion
This will log the rendered JSX without the `style` prop.
```

mdjastrzebski commented on website/docs/API.md:
```diff
@@ -735,6 +755,9 @@ function configure(options: Partial<Config>)  {}
 
+#### `debugOptions` option
+
+Default debugOptions to be used when calling debug. These default options will be overridden by the ones you specify directly when calling debug
```
```suggestion
Default `debugOptions` to be used when calling `debug()`. These default options will be overridden by the ones you specify directly when calling `debug()`.
```

mdjastrzebski reviewed (commented):
I've done initial review on implementation and left some comments there. Overall I like the code so far, looks pretty solid. 

The point I want to rise is the desired API we want to have. 

The current API is:
```
export type FormatOptions = {
  filterProps?: (
    propName: string,
    propValue: unknown,
    node: ReactTestRendererJSON
  ) => boolean;
};
```
While it's quite flexible, I think we could investigate alternative approaches, based on the use cases we want to support.

### Idea 1: allowlist/blocklist
As far as I understand the primary use case is to print only some relevant props and/or exclude some props that are cluttering the output. We are rarely interested in inspecting the prop value. We could achieve that by somewhat easier API:

```ts
type FormatOptions = {
  allowedProps?: string[];
  ignoredProps?: string[];
}
```

If specified `allowedProps` will contain the names props that will be printed, no other props would be printed by `format`. Otherwise, all props will be printed except the ones specified in `ignoredProps`.

Pros: simpler API for majority use case
Cons: less flexible

### Idea 2: ability to transform props
Opposite to idea 1, this makes the API more flexible by allow prop transformation before printing. This could allow for both removal of props that are not interesting to as, or applying some transformation to their value that would improve the readability:

Version 1 - per prop
```ts
type FormatOptions = {
  mapProp?: (
    propName: string,
    propValue: unknown,
    node: ReactTestRendererJSON
  ) => [name, value];
}
```

Version 2 - per node
```ts
type FormatOptions = {
  mapProps?: (props: Record<string, unknown>) => Record<string, unknown>; // output props to be displayed
}
```

### Examples
Original:
```ts
debug(<JSX />, { filterProps: (name, value, node) => name === 'style' });
debug(<JSX />, { filterProps: (name, value, node) => ['style', 'testID'].includes(name) });
```

Idea 1:
```ts
debug(<JSX />, { allowedProps: ['style'] });
debug(<JSX />, { allowedProps: ['style', 'testID'] });
```

Idea 2 - per prop:
```ts
debug(<JSX />, { mapProp: (name, value, node) => name === 'style' ? [name, value] : undefined });
debug(<JSX />, { mapProp: (name, value, node) => ['style', 'testID'].includes(name ? [name, value] : undefined });
```

Idea 2 - per node:
```ts
debug(<JSX />, { mapProps: ({ style }) => ({ style }) });
debug(<JSX />, { mapProps: ({ style, testID }) => ({ style, testID }) });
```

@pierrezimmermannbam could you provide typical examples you envision this option will handle so that we can figure out which API options is the best to address it.

AugustinLF commented on src/__tests__/__snapshots__/render.test.tsx.snap:
```diff
@@ -124,6 +124,224 @@ exports[`debug changing component: bananaFresh button message should now be "fre
+exports[`debug with all props filtered 1`] = `
+"<View>
+  <Text>
+    Is the banana fresh?
```
Shouldn't `children` be filtered also?

mdjastrzebski replied:
I think this is due to `children` prop being extracted from `props` in both TestInstance and JSON. Imo we should keep this as is, as a reasonable behaviour for most cases.

pierrezimmermannbam replied:
I was surprised by this as well but I actually think that this is fine, we'll want to keep the children prop in most cases and not having to explicitly include it is nice

AugustinLF reviewed (approved):
~Looks good to me, my feedback has been addressed, I'll let you address @mdjastrzebski comments~

AugustinLF reviewed (commented):
Good for me, I'll let you address @mdjastrzebski's comment meanwhile. I won't hit the `Approve` button to be sure it doesn't get merged by mistake :)

pierrezimmermannbam commented:
@mdjastrzebski you're right we should consider alternative approaches. I plan to use this feature in the following ways : 

- remove globally some props that aren't useful for debugging with configure api (style or path for instance)
- create custom debug utils that would be used for debugging each query using screen (for instance debugText would filter all props except children)
- use a more custom version in tests if the need arises but this should be pretty rare

Overall I'd mostly use this either with configure or have a set of utils that I would write only once so I like the idea of having a rather flexible api. However It should be accessible to users too so maybe we could document it a bit more with some use cases that we find relevant.

AllowedProps and ignoredProps is probably simpler than filterProps so I think it's interesting. I'm  not sure however that it's that much simpler, plus i'm afraid that there may be some conflicts between ignored and allowed props especially if one has been set through configure. Also i was thinking of globally filtering props whose value is a function which you cannot do without access to the propValue. I also like to have a single option for filtering props, because we may add other options later 
and the more there are the more it's going to be complicated to understand all of them.

mapProps is very interesting though, being able to flatten styles in the debug function for instance could be a very useful option. Are there any other use cases though ? Maybe directly having a flattenStyle option would be more convenient if that's not the case. What I'm afraid of is that it seems less obvious to me why you'd want to map props than why you'd want to filter props but it would indeed allow way more flexibility.  I think I wouldn't mind having both the filterProps and the mapProps option though. 

What do you think would be the best api based on the examples I provided ?

mdjastrzebski commented:
@pierrezimmermannbam not sure if you're participating in Hacktoberfest but if so I've added `hacktopberfest-accept`, I plan to review this again in the following days.

mdjastrzebski commented:
@pierrezimmermannbam 

Imo based on your use cases I think `mapProps` (per node version) makes most sense and is also the most flexible approach. It allows for both filtering & transforming props (e.g. flatten styles), while also allowing for basic output on a prop combination, e.g. output `style` prop only if `disabled` props is not set, etc. I'm curious about @AugustinLF take on the API.

re accept list/ignore list priority: I think a reasonable assumption is that `accept` list takes priority over `ignore` list. If you have `accept` list then you do not use `ignore` list at all.

re ease of use: we should have some documentation examples for common use cases, that should clarify the usage.

re having both `filterProp` and `mapProps` that would make the API a bit confusing as we would have two functions doing very similar things, and also it's not clear how should they interact. Which should be used if both are present, etc

I would be wary of introducing niche options like `flattenStyles` as it can be done with `mapProps` and imo is not worth to complicating our API.

AugustinLF commented:
I think it's fair to go with `mapProps`. It is an option for advanced users, so we should aim it to make it complete. And yes, for things like `flattenStyles` we could offer it as a recipe. It doesn't have to be in our API. Or if highly requested, we could expose it as an export, that could then be passed to the `mapProps`.

pierrezimmermannbam commented:
I also agree, I'll change the api and add some examples of usage in the documentation

pierrezimmermannbam commented:
@mdjastrzebski  I changed the api to use mapProps instead and updated the website docs to include more examples of the option usage. I also treated the comments you left. 

There is one last thing I'm not sure about though, should we keep the compatibility with previous versions for debugOptions ? The debug util is not meant to stay in the code, it's just a debugging tool so there should be little to no issues regarding the migration. I feel like it adds unnecessary complexity in the RNTL codebase while providing very little value to users. I think this could be merged as is and be included in a release before next major version but I would remove retro compatibility with the next major version, wdyt ?

mdjastrzebski commented on src/render.tsx:
```diff
@@ -135,18 +136,31 @@ function updateWithAct(
+    if (typeof options === 'string') {
+      // eslint-disable-next-line no-console
+      console.warn(
+        'Please use debug({message; "message"}) instead of debug("message")'
```
```suggestion
        'Using debug("message") is deprecated and will be removed in future release, please use debug({ message; "message" }) instead.'
```

mdjastrzebski commented on website/docs/API.md:
```diff
@@ -737,6 +782,9 @@ function configure(options: Partial<Config>)  {}
 
+#### `defaultDebugOptions` option
+
+Default `debugOptions` to be used when calling `debug()`. These default options will be overridden by the ones you specify directly when calling `debug()`. Can be used namely to transform prop values for more readability or filter out props that have little value when debugging tests (e.g. svg's path prop).
```
I would rather refer user to `debug` `options` docs for details rather than repeating the use cases.

```suggestion
Default `debugOptions` to be used when calling `debug()`. These default options will be overridden by the ones you specify directly when calling `debug()`. 
```

pierrezimmermannbam replied:
agreed, I added a link to debug options doc

mdjastrzebski commented on website/docs/API.md:
```diff
@@ -223,6 +250,24 @@ Hold the value of latest render call for easier access to query and other functi
 
 Its value is automatically cleared after each test by calling [`cleanup`](#cleanup). If no `render` call has been made in a given test then it holds a special object that implements `RenderResult` but throws a helpful error on each property and method access.
 
+This can also be used to build test utils that would normally require to be in render scope, either in a test file or globally for your project. For instance in a test file where we fill a form in multiple tests: 
```
The concept of building test utils seems like a useful idea, however the `typeEmail` example seems quite artificial and I would not recommend using such util over explicit `fireEvent` + query.

@pierrezimmermannbam do you have idea for something that we could actually recommend to your users here?

pierrezimmermannbam replied:
This is something is sometimes do when I test forms and have a lot of tests, imo it makes tests more readable and very easy to understand while not making them harder to debug since you still have the same error messages. It's even more interesting when it allows to hide testID usage since the queries by testID do not really help to understand the test. An example of this is waiting for a loader to disapear, which is often done in every test when testing at page level

```tsx
const waitForLoaderToDisapear = async () => {
   await waitForElementToBeRemoved(() => screen.getByTestID('PageLoader'));
}
```

Does this example make more sense to you ?

mdjastrzebski commented on website/docs/API.md:
````diff
@@ -193,6 +202,24 @@ optional message
 </View>
 ```
 
+You can use the mapProps option to transform the props that will be printed : 
````
```suggestion
You can use the `mapProps` option to transform the props that will be printed : 
```

mdjastrzebski commented on website/docs/API.md:
````diff
@@ -193,6 +202,24 @@ optional message
+You can use the mapProps option to transform the props that will be printed : 
+
+```jsx
+render(<View style={{backgroundColor: 'red'}}/>);
+debug({ mapProps : ({style, ...props}) => ({props}) })
````
formatting:
```suggestion
render(<View style={{ backgroundColor: 'red' }}/>);
debug({ mapProps : ({style, ...props}) => ({ props }) })
```

mdjastrzebski commented on website/docs/API.md:
```diff
@@ -193,6 +202,24 @@ optional message
+
+This will log the rendered JSX without the `style` props. 
+
+This option can be used to target specfic props when debugging a query (for instance keeping only children prop when debugging a byText query) and you can also transform prop values so that they are more readable (e.g. flatten styles).
```
Not sure about mentioning `children` prop in this case, as it is treated specially by `mapProps`

```suggestion
This option can be used to target specific props when debugging a query (for instance keeping only `children` prop when debugging a `getByText` query) and you can also transform prop values so that they are more readable (e.g. flatten styles).
```

pierrezimmermannbam replied:
I changed the order so that it is explained first that children prop cannot be filtered then give the example of usage, since I believe this will be one of the most common, at least until byRole queries become more popular

mdjastrzebski commented on website/docs/API.md:
```diff
@@ -193,6 +202,24 @@ optional message
+
+This option can be used to target specfic props when debugging a query (for instance keeping only children prop when debugging a byText query) and you can also transform prop values so that they are more readable (e.g. flatten styles).
+
+The `children` prop cannot be filtered so the following will print all rendered components with all props but children filtered.
```
```suggestion
The `children` prop cannot be filtered out so the following will print all rendered components with all props but children filtered.
```

mdjastrzebski commented on website/docs/API.md:
```diff
@@ -170,15 +171,23 @@ Usually you should not need to call `unmount` as it is done automatically if you
 
-Pretty prints deeply rendered component passed to `render` with optional message on top.
+Pretty prints deeply rendered component passed to `render`. 
+
```
Pls add lower-level headers for options as we do for other APIs

```suggestion
#### `message` option
```

pierrezimmermannbam replied:
Very good idea I hadn't noticed this was done for other apis !

mdjastrzebski commented on website/docs/API.md:
````diff
@@ -193,6 +202,24 @@ optional message
 </View>
 ```
 
````
```suggestion
#### `mapProps` option
```

mdjastrzebski commented on typings/index.flow.js:
```diff
@@ -413,6 +426,7 @@ declare module '@testing-library/react-native' {
 
   declare interface Config {
     asyncUtilTimeout: number;
+    debugOptions?: $Shape<DebugOptions>;
```
```suggestion
    defaultDebugOptions?: $Shape<DebugOptions>;
```

pierrezimmermannbam replied:
Oh good catch, I keep missing the flow types updates

mdjastrzebski commented on typings/index.flow.js:
```diff
@@ -349,8 +349,21 @@ interface Thenable {
+  node: ReactTestRendererJSON
+) => { [string]: mixed };
+
+type FormatOptions = {
+  mapProps?: MapPropsFunction,
+};
+
+type DebugOptions = {
+  message?: string,
+} & FormatOptions;
```
`FormatOptions` does not seem to be used anywhere, let's merge it with `DebugOptions` for flow


```suggestion
type DebugOptions = {
  message?: string,
  mapProps?: MapPropsFunction,
};
```

mdjastrzebski commented on src/render.tsx:
```diff
@@ -135,18 +136,31 @@ function updateWithAct(
-  function debugImpl(message?: string) {
+  function debugImpl(options?: DebugOptions | string) {
+    const { defaultDebugOptions } = getConfig();
     const json = renderer.toJSON();
```
style, nit:
Pls move `const json = ...` line just above the `if (json)` statement, as it is not used during option calculation part.

mdjastrzebski commented on src/helpers/format.ts:
```diff
@@ -1,10 +1,53 @@
+) =>
   prettyFormat(input, {
-    plugins: [plugins.ReactTestComponent, plugins.ReactElement],
+    plugins: [getCustomPlugin(options.mapProps)],
```
I think it's still a good idea to make `getCustomPlugin` tweak only `ReactTestComponent` while passing `ReactElement` normally
```suggestion
    plugins: [getTestComponentPlugin(options.mapProps), plugins.ReactElement],
```

pierrezimmermannbam replied:
Absolutely, it also makes the code way simpler !

mdjastrzebski commented on src/helpers/debugDeep.ts:
```diff
@@ -1,18 +1,27 @@
+  const message = typeof options === 'string' ? options : options?.message;
+
+  const formatOptions =
+    typeof options === 'object' ? { mapProps: options.mapProps } : undefined;
```
Should prevent crash if user passes `null`, because `typeof null === 'object'`

```suggestion
    typeof options === 'object' ? { mapProps: options?.mapProps } : undefined;
```

mdjastrzebski commented on src/__tests__/render.test.tsx:
```diff
@@ -191,11 +211,121 @@ test('debug changing component', () => {
+
+  const { debug, getByText } = render(<Banana />);
+
+  debug({ mapProps: ({ children }) => ({ children }) });
```
This test seems confusing, as `children` in not included in the props. Pls adjust or remove the test.

pierrezimmermannbam replied:
Oh you're right, I had written them before I realized the children prop wasn't filtered and it does not really make sense anymore. I deleted one test and modified the other one

mdjastrzebski commented on src/__tests__/render.test.tsx:
```diff
@@ -191,11 +211,121 @@ test('debug changing component', () => {
+
+  expect(stripAnsi(mockCalls[0][0])).toMatchSnapshot();
+
+  expect(getByText('Change freshness!')).toBeTruthy();
```
Having `getByText` assertion does not seem to contribute to the test.
```suggestion
```

pierrezimmermannbam replied:
Actually what I was trying to do here was make sure that filtering out props wouldn't really delete them but since the children prop is never filtered it doesn't make sense. I'll replace this by a more explicit and relevant test

mdjastrzebski commented on src/__tests__/render.test.tsx:
```diff
@@ -191,11 +211,121 @@ test('debug changing component', () => {
+
+  expect(stripAnsi(mockCalls[0][0])).toMatchSnapshot();
+
+  expect(getByText('Change freshness!')).toBeTruthy();
```
Having `getByText` assertion does not seem to contribute to the test.
```suggestion
```

mdjastrzebski commented on src/__tests__/render.test.tsx:
```diff
@@ -191,11 +211,121 @@ test('debug changing component', () => {
+
+  expect(stripAnsi(mockCalls[0][0])).toMatchSnapshot();
+
+  expect(getByText('Change freshness!')).toBeTruthy();
```
Having `getByText` assertion does not seem to contribute to the test.
```suggestion
```

mdjastrzebski commented on src/__tests__/render.test.tsx:
```diff
@@ -191,11 +211,121 @@ test('debug changing component', () => {
   const mockCalls = (console.log as any as ConsoleLogMock).mock.calls;
 
-  expect(stripAnsi(mockCalls[4][0])).toMatchSnapshot(
+  expect(stripAnsi(mockCalls[0][0])).toMatchSnapshot(
```
What caused the index change here?

pierrezimmermannbam replied:
The fact that mocks are being cleared automatically do to the change I made in jest.config.js. We were checking the 4th call because it was called in previous test but now mock calls are reset before each test

mdjastrzebski commented on src/helpers/format.ts:
```diff
@@ -1,10 +1,53 @@
+    serialize: (val, config, indentation, depth, refs, printer) => {
+      if (plugins.ReactTestComponent.test(val)) {
+        if (mapProps && val.props) {
+          val.props = mapProps(val.props, val);
```
Not sure if it's a good idea to mutate the passed `val` object. Can we rather use object spread operator to create a modified copy?

pierrezimmermannbam replied:
I had my doubt on this, this is why I added a test to verify that the original object is not modified and it's not the case so I think it's fine as is. Also I believe that using the spread operator would create a shallow copy so it would not really be a protection against modifications either

mdjastrzebski replied:
format works on JSON output not ReactTestInstances, but still invoking `format` would mutate the JSON output. I've changed that to object copy, not mutation.

mdjastrzebski reviewed (changes requested):
@pierrezimmermannbam This PR looks really solid. I've added some suggestions for finishing touches and edge cases. 

Pls add some real-life examples to the docs on how users could you this to improve their debugging output.

pierrezimmermannbam commented:
@mdjastrzebski  I think I answered all your comments. I added an example to the doc and detailed more the flatten style example, do you think of anything else ?

mdjastrzebski reviewed (approved):
Looks great! Good work @pierrezimmermannbam 🚀 

I've made some cosmetic tweaks to docs and code before merging.
